# Logistic Regression in SciKit Learn - Lab

## Introduction 

In this lab, we are going to fit a logistic regression model to a dataset concerning heart disease. Whether or not a patient has heart disease is indicated in the final column labelled 'target'. 1 is for positive for heart disease while 0 indicates no heart disease.

## Objectives
You will be able to:

* Understand and implement logistic regression
* Compare testing and training errors

## Let's get started!

In [1]:
#Starter Code
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
import pandas as pd

In [2]:
#Starter Code
df = pd.read_csv('heart.csv')
df.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

## Define appropriate X and y
Recall the dataset is whether or not a patient has heart disease and is indicated in the final column labelled 'target'. With that, define appropriate X and y in order to model whether or not a patient has heart disease.

In [8]:
#Your code here 
y = df.target
X = df.drop(columns=['target'])
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

## Normalize the Data
Normalize the data prior to fitting the model.

In [9]:
for col in x.columns:
    x[col] = (x[col] - min(x[col])) / (max(x[col] - min(x[col])))
X.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

## Train Test Split
Split the data into train and test sets.

In [10]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.2, random_state=123)

## Fit a model
Fit an intial model to the training set. In SciKit Learn you do this by first creating an instance of the regression class. From there, then use the **fit** method from your class instance to fit a model to the training data.

In [11]:
logreg = LogisticRegression(fit_intercept = False, C = 1e12) #Starter code
model = logreg.fit(X_train, y_train)
model

/anaconda3/lib/python3.7/site-packages/sklearn/linear_model/logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


LogisticRegression(C=1000000000000.0, class_weight=None, dual=False,
          fit_intercept=False, intercept_scaling=1, max_iter=100,
          multi_class='warn', n_jobs=None, penalty='l2', random_state=None,
          solver='warn', tol=0.0001, verbose=0, warm_start=False)

## Predict
Generate predictions for the train and test sets. Use the **predict** method from the logreg object.

In [12]:
yhat_train = model.predict(X_train)
yhat_test = model.predict(X_test)

## Initial Evaluation
How many times was the classifier correct for the training set?

In [24]:
import numpy as np

train_residuals = np.abs(yhat_train - y_train)
print(pd.Series(train_residuals).value_counts())
print(pd.Series(train_residuals).value_counts(normalize=True))

0    216
1     26
Name: target, dtype: int64
0    0.892562
1    0.107438
Name: target, dtype: float64


## How many times was the classifier correct for the test set?

In [26]:
test_residuals = np.abs(yhat_test - y_test)
print(pd.Series(test_residuals).value_counts())
print(pd.Series(test_residuals).value_counts(normalize=True))

0    47
1    14
Name: target, dtype: int64
0    0.770492
1    0.229508
Name: target, dtype: float64


## Analysis
Describe how well you think this initial model is based on the train and test performance. Within your description, make note of how you evaluated perforamnce as compared to our previous work with regression.

Given a 77% accuracy rate with the test data, as compared to 89% for the train data suggests that this model is likely a bit overfitted.

In addition, we don't have enough information here to properly assess the strenth of the model. Specifically, we don't know the makeup of false positives and false negatives. Given that we are testing for heart disease, we would like to minimize the number of false negatives.

## Summary

In this lab, you practiced a standard data science pipeline, importing data, splitting into train and test sets and fitting a logistic regression model. In the upcoming labs and lessons, we'll continue to investigate how to analyze and tune these models for various scenarios.